### DISCUSSION TITLE
relationship between real dimensions and dimensions in Unity

### DISCUSSION THREAD

**Discussion by G-P-x** [mlavik1/UnityVolumeRendering#197] [Q&A]
Hi, after loading the object by importing a folder with .dcm files, I wanted to try to obtain the volume of the object. So I was wondering if the size of the object in unity is the same as it is in reality or is it scaled somehow.

**Comment by mlavik1** [comment]
Hi!
By default, the volume size will be normalised to fit inside a small box. However, if you se the scale of the outer GameObject ("VolumeRenderedObject") to (1,1,1) then the scale should now be in real-world metric scale. 
This is new, so I haven't had time to update the documentation yet. Also it's only available in the current master branch - not in the latest release. Planning to release a new version soon!

I should probably add an importer setting to optionally make it always use real-world metric scale. The normalisation was originally done since some dataset formats (RAW) might not contain info about the scale units.

**Reply by G-P-x** [reply]
I understand,
the problem is that the way i calculate the volume is by comparing the object with imported prefabs like cube and sphere.
I would like the object shown, let's assume a head, to have the same dimensions as it would in reality.
For example i imported this dicom dataset  ( https://unicadrsi-my.sharepoint.com/:i:/g/personal/g_pau15_studenti_unica_it/Efvoabbix99Gu3EgdSaNJ80Bp2iIZpzloxraIxlci7J7tw?e=VyiYDS ); as you can see the z dimension is 0.4m while it should be around 0.1m according to metadata. Is there another way to adjust it? changing the scale of the outer GameObject to (1,1,1)  doesn't solve the problem.

**Reply by mlavik1** [reply]
Hmm.. So I assume the Z scale of the "VolumeContainer" GameObject shows as 0.4 then, right?

A few questions:
- What is the pixel spacing, slice thickness and dimension of that dataset?
- Are you using the SimpleITK integration, or the built-in DICOM importer?
- Can I download this dataset somewhere? (or another dataset where you see the same issue).
- You're testing with the latest master branch, right?

I calculate the size as `spacing * dimension`, (by "dimension" I mean the number of rows/columns/slices) which seems to look correct on the datasets I've tested with so far. Is this how you calculated your expected size as well? (when you said it should be 0.1mm)
You can see the calculation in `SimpleITKImageSequenceImporter.cs`  I wonder if I'm missing something though..

Example dataset:
Dimension: 512, 512, 241 (512 rows, 512 columns, 241 slices)
Pixel Spacing: 0.644531, 0.644531, 1.25
=>
Size: 330m, 330mm, 301mm

**Reply by G-P-x** [reply]
okay, maybe i found the solution, i changed the scale in the script VolumeObjectFactory.cs
if (dataset.scaleX != 0.0f && dataset.scaleY != 0.0f && dataset.scaleZ != 0.0f)
            {
                //float maxScale = Mathf.Max(dataset.scaleX, dataset.scaleY, dataset.scaleZ);
                volObj.transform.localScale = new Vector3(dataset.scaleX , dataset.scaleY , dataset.scaleZ ) * 0.001f;
            }
instead of
            //if (dataset.scaleX != 0.0f && dataset.scaleY != 0.0f && dataset.scaleZ != 0.0f)
            //{
            //    float maxScale = Mathf.Max(dataset.scaleX, dataset.scaleY, dataset.scaleZ);
            //    volObj.transform.localScale = new Vector3(dataset.scaleX / maxScale, dataset.scaleY / maxScale, dataset.scaleZ / maxScale);
            //}, 

Using this new code, the Z scale of the external object is 0.09m which is the same dimension it have in real. The scale of the inner object remain the same (1,1,1).

https://unicadrsi-my.sharepoint.com/:i:/g/personal/g_pau15_studenti_unica_it/ET3ib1zY9eFGhau7I69PUIABCkpApIgrIf7Hlsc2gSy4_w?e=vdzdh0

https://unicadrsi-my.sharepoint.com/:i:/g/personal/g_pau15_studenti_unica_it/EXXD6P1vHIFLp44okViNv5oB2zEyCgwsD_PKuDCR_BYSYA?e=a1wIxY

https://unicadrsi-my.sharepoint.com/:i:/g/personal/g_pau15_studenti_unica_it/EbN_pmqCI5hDuRSQkwdQyYcBDZRhtl1Y0J92xsQRe9T3TA?e=oxF8Sj

I'm creating a demo, so i'm testing it with just two dataset, but for both this correction work

Anyway, this is the dataset i'm using : 
1. Circle of Wills: https://unicadrsi-my.sharepoint.com/:f:/g/personal/g_pau15_studenti_unica_it/EpXxDovKdvJNrKfjOPbZXJQB4eErn9vibYkeROTmvOuy7w?e=0rgdBD

As you can see from the metadata
dimension : 512 * 512 * 137
PixelSpacing : 0.44921875, 0.44921875, 0.7 => Size 230 mm , 230 mm , 95 mm 

0.7 is the SpacingBetweenSlices while SliceThickness is 1.4. I'm not sure what value you refer. Then also in 3D Slicer, the dimesion of the object is arounf 97 mm for the Z-axis.

Another thing that maybe could be important is the SliceLocation, 0 for the first and 97.3 for the last. 
I tried with SimpleITK enable and disable, but for me didn't change anything

**Comment by mlavik1** [comment]
Thanks for sharing the dataset! I'll have a look at it tomorrow. Talk to you soon!

**Comment by mlavik1** [comment]
Hmm.. with that change you added the imported dataset becomes so small that I can't see it.. Are we on the same branch? (latest "master" branch)

But when I import the dataset without any code changes (on latest master branch) and set the scale of the outer object to (1,1,1) the size appears to be (230 mm, 230 mm, 95 mm) as expected.
To compare: If I create a box (which has a size of 1 (1000 mm), and I set its size to (0.23, 0.095, 0.23) (same as above, but in metre) their sizes will be the same:
![image](https://github.com/mlavik1/UnityVolumeRendering/assets/6906872/82edb1ef-3ce9-4e7d-876a-efbde14c8bdb)

How are you measuring the sizes on your side?

By the way, this is the object I'm refering to when I say "outer object":
![image](https://github.com/mlavik1/UnityVolumeRendering/assets/6906872/a56a81de-5ed0-4529-87dc-ea196ffbd23c)

Setting it to (1,1,1) should do the same as if you removed that line of code that normalises the size in `VolumeObjectFactory.CreateObjectInternal`

**Comment by G-P-x** [comment]
Hi, i apologize for not responding earlier.
For my application, i need  the object to be displayed in real size, so i measured it using Unity's ProBuilder tool.
Now it occurred to me tath, as you rightly pointed out, setting the scale of  _VolumeRenderedObject_1-001.dcm_ to  (1,1,1) results in obtaining the correct measurement using your size measurement tool. However, the object in Unity scene is obviously distorted. That's a problem for me because the way i estimate the volume is by overlaying simple prefabs figures such as cubes and sphere onto the object in the scene. If the object in the scene is distorted, using the tool you created provides correct measurements, whereas when i do it myself, i get inaccurate measurements.

**Reply by mlavik1** [reply]
Hmm.. how does the object look distorted? Is the scale wrong, or is it there an issue with the actual volume rendering? Could you share some pictures for comparison maybe? That would be very helpful :)

I'm going away for a few days, but I'll have a look at it again when I am back, and try compare the dataset you sent me to how it looks in some other viewers (such as 3D Slicer).

Oh and just to check, are you on Windows or something else? And are you using default legacy renderer with D3D11?

Have a great weekend!
